text/3503-frontmatter.md

```diff
@@ -0,0 +1,512 @@
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code for quality of error messages, etc
+- Leave the door open in case we want to reuse the syntax for embedded lockfiles
+- Leave the door open for single-file `lib`s
+
+This proposed syntax builds off of the precedence of Rust syntax being specialized for an external tool
+(doc-comments for rustdoc).
+However, a new syntax is used instead of extending the comment syntax:
+- Focused on being parsed by arbitrary tools (cargo, vim plugins, etc) without understanding the full Rust grammar
+- Side steps compatibility issues with both
+  user expectations with the looseness of comment syntax (making it harder to parse)
+  and any existing comments that may look like a new structured comment syntax
+
+Benefits for the proposed solution are:
+- Visually/syntactically lightweight
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+  - Cargo needs a simple syntax subset to parse to ensure it can adequately report errors for the part it cares about but does not get in the way of rustc reporting the high quality errors it cares about
+- In the future, this can be leveraged by other build systems or tools
+
+### Alternative 1: Doc-comment
+
+```rust
+#!/usr/bin/env cargo
+
+//! ```cargo
+//! [package]
+//! edition = "2018"
+//! ```
+
+fn main() {
+}
+```
+
+Benefits
+- Parsers are available to make this work (e.g. `syn`, `pulldown-cmark`)
+- Familiar syntax both to read and write.
+  - When discussing with a Rust author, it was pointed out many times people preface code with a comment specifying the dependencies ([example](https://github.com/dtolnay/prettyplease#example)), this is the same idea but reusable by cargo
+  - When discussing on forums, people expressed how they had never seen the syntax but instantly were able to understand it
+
+Downsides:
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Having the explain the overloading of concepts to new users
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- Might be a bit complicated to do edits (translating between location within
+  `toml_edit` spans to the location within `syn` spans especially with different comment styles)
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+- Requires pulling in a full markdown parser to extract the manifest
+  - Incorrectly formatted markdown would lead to a missing manifest and confusing error messages at best or silent incorrect behavior at worse
```

Comment:
I'm not entirely convinced by this. Extracting the manifest from the comment does not seem like it would need a fullblown markdown parser, and if we see a module-level doc-comment but no valid manifest block it should be possible to show a warning to the user.

Reply:
> Extracting the manifest from the comment does not seem like it would need a fullblown markdown parser

We could go to the commonmark spec and hand write a subset of a markdown parser just to deal with code fences.  [I've done that before](https://github.com/assert-rs/trycmd/blob/main/src/schema.rs#L192) and just through this RFC I've found I have bugs in it. Having two separate parsers means you have two different semantics you are dealing with.  This was a big reason why `cargo add` took so long to be merged: the cargo team wanted only a single set of semantics for TOML while not giving up format preserving.